### PR TITLE
Named mesh constructors and split any mesh into all simplex mesh [tri-tet-lor]

### PR DESCRIPTION
**Pull request by pazner**
The main functionality of this PR is to split any mesh into an all-simplex mesh according to [this algorithm](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.35.4908). This is particularly useful for LOR preconditioners, which are both sparser and give better conditioning when assembled on simplex rather than tensor-product meshes.

Along the way, I added a move constructor and several named constructors for the `Mesh` and `ParMesh` classes. The motivation for this was that I didn't want to add yet-another overloaded mesh constructor, and so (after some discussions with @v-dobrev) decided to implement this as a "named constructor".

I am probably not the only one who finds this situation confusing. Before:
```c++
// Given Mesh mesh...
Mesh m1(mesh, 1); // Copy constructor
Mesh m2(&mesh, 1, 1); // LOR constructor
Mesh m3(1, 1); // 1D grid
Mesh m4(1, 1, 1); // "Init"-style constructor
Mesh m5(1, 1, eltype); // 2D grid
Mesh m6(1, 1, 1, eltype); // 3D grid
```

After:
```c++
// Given Mesh mesh...
Mesh m1(mesh, 1); // Copy constructor
Mesh m2 = Mesh::MakeRefined(mesh, 1, 1); // LOR constructor
Mesh m3 = Mesh::MakeCartesian1D(1, 1); // 1D grid
Mesh m4(1, 1, 1); // "Init"-style constructor
Mesh m5 = Mesh::MakeCartesian2D(1, 1, eltype); // 2D grid
Mesh m6 = Mesh::MakeCartesian3D(1, 1, 1, eltype); // 3D grid
Mesh m7 = Mesh::MakeSimplicial(mesh); // New simplicial constructor
```

⚠️ One consequence of this change is that constructing a new `Mesh` pointer with the named constructors is somewhat more verbose, e.g. `Mesh *mesh = new Mesh(Mesh::MakeCartesian1D(…));`.

❗ Please leave feedback if you would prefer to see any changes to this PR.

Since the copy constructor of `Mesh` is marked `explicit`, in order to return a `Mesh` from a function (as is needed by the named constructors), the move constructor is required. **Note:** in practice, the move constructor is not actually called — the compiler is able to entirely elide all extraneous calls to the constructors. In any case, the move constructor should be quite cheap since it just swaps pointers.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2016](https://github.com/mfem/mfem/pull/2016) | @pazner | @v-dobrev | @barker29 + @jakubcerveny | 02/02/21 | 03/09/21 | 03/31/21 | |

### CODE REVIEW

**Comment by pazner** [comment]
@v-dobrev, can you take a look at this/assign reviewers?

**Review by barker29 — CHANGES_REQUESTED** [review]
This is a very nice feature, but it needs just a bit of polishing before merging. In particular:

1. It needs some unit testing of the parallel algorithm. The handling of `Group`s is non-trivial and I would be more confident in this code if it were unit tested.
2. This fails on periodic meshes, which is probably fine, but the error message is not informative (ideally it would say something like "MakeSimplicial() doesn't work on periodic meshes.").
3. Similarly, it fails on the AMR meshes `amr-hex.mesh`, `amr-quad.mesh`, which again is fine but it would be nice if the error message could be informative.
4. It appears to run on NURBS meshes but it is not at all clear what it is actually doing - probably nothing? It's a bit weird to run something called `MakeSimplicial` on a NURBS mesh and have it just run.

File: mesh/mesh.hpp
Comment (pazner) [reply mesh/mesh.hpp]: Done

File: mesh/mesh.hpp
Comment (pazner) [reply mesh/mesh.hpp]: Fixed

File: mesh/mesh.hpp
Comment (pazner) [reply mesh/mesh.hpp]: Fixed

File: mesh/mesh.cpp
Comment (pazner) [reply mesh/mesh.cpp:3252]: Making this change is fine by me. Other opinions, @v-dobrev, @jakubcerveny?

File: mesh/mesh.hpp
Comment (pazner) [reply mesh/mesh.hpp:716]: Making them private might be too breaking, but marking them as deprecated could be a good compromise.

They are used somewhat extensively in the unit tests, so it will require changing a lot of code. Perhaps it's better to do it in a separate PR.

**Comment by pazner** [comment]
> This is a very nice feature, but it needs just a bit of polishing before merging. In particular:

Thanks for the good feedback.

> 1. It needs some unit testing of the parallel algorithm. The handling of `Group`s is non-trivial and I would be more confident in this code if it were unit tested.

That's a good suggestion, and I agree that the handling of the groups is complicated. I added a test that solves a p=1 Poisson problem on the hex and tet meshes with exact solution that lies in the P1 space (and so also in the Q1 space), and check that the resulting solution is the same up to solver tolerance. It's not really a "unit test" (I think doing true unit testing of the groups would be quite challenging), but it should be a decent stress test that the resulting mesh really is a valid parallel mesh. If you have ideas for how to do good unit testing for parallel meshes, I'd be interested in hearing.

> 2. This fails on periodic meshes, which is probably fine, but the error message is not informative (ideally it would say something like "MakeSimplicial() doesn't work on periodic meshes.").

Handling of the "nodes" is tricky, since there is not a natural injection from the nodal FESpace defined on the original mesh to the nodal FESpace defined on the simplex mesh. p=1 is a special case since the vertices are preserved by the new mesh, and this would in principle mean that some support for straight-sided periodic meshes is possible. Some of the checks (e.g. in `CheckElementOrientation` and `FinalizeTopology`) are failing for some of the periodic meshes, so getting this right may not be straightforward.

It's not so easy to error on a periodic mesh, because there is no easy way of detecting one. Looking for an L2 nodal space would be one way, but not every mesh with L2 nodes is periodic…

> 3. Similarly, it fails on the AMR meshes `amr-hex.mesh`, `amr-quad.mesh`, which again is fine but it would be nice if the error message could be informative.

Yeah, I think for now it just makes sense to fail on AMR meshes. I improved the error message in this case.

> 4. It appears to run on NURBS meshes but it is not at all clear what it is actually doing - probably nothing? It's a bit weird to run something called `MakeSimplicial` on a NURBS mesh and have it just run.

I think in this case it's just treating the mesh as straight-sided, and ignoring the curvature (related to 2 above). I could add an error if it's given a NURBS mesh if you think that's better.

File: tests/unit/mesh/test_pmesh.cpp
Comment (tzanio) [inline tests/unit/mesh/test_pmesh.cpp:170]: Travis [doesn't like](https://travis-ci.com/github/mfem/mfem/jobs/482864299#L950) this semicolon.

File: mesh/mesh.cpp
Comment (jakubcerveny) [inline mesh/mesh.cpp:4603]: There already is a very similar method one page below, is the new one necessary?

File: mesh/mesh.hpp
Comment (jakubcerveny) [inline mesh/mesh.hpp:732]: Maybe change comment to "Deprecated: see @a MakeCartesian1D"?

**Review by jakubcerveny — APPROVED** [review]
Nice, approved modulo small comments below.

File: mesh/mesh.hpp
Comment (pazner) [reply mesh/mesh.hpp]: Done.

File: mesh/mesh.cpp
Comment (pazner) [reply mesh/mesh.cpp]: Good idea, I changed how it works now.

File: mesh/mesh.cpp
Comment (pazner) [reply mesh/mesh.cpp]: I think that's a good idea. I implemented it.

File: mesh/mesh.cpp
Comment (pazner) [reply mesh/mesh.cpp]: Ok, I did it

File: mesh/mesh.hpp
Comment (pazner) [reply mesh/mesh.hpp:718]: I implemented these changes in #2054.

File: mesh/mesh.cpp
Comment (pazner) [reply mesh/mesh.cpp:4028]: Fixed, thanks.

File: mesh/mesh.cpp
Comment (pazner) [reply mesh/mesh.cpp]: No, it's not necessary. I removed it.

**Comment by pazner** [comment]
Check out #2054 to see the changes that would be required if we want to deprecate the old-style mesh constructors. Let me know if you think we should go ahead with that @barker29 @jakubcerveny @v-dobrev

**Review by barker29 — APPROVED** [review]
I think this is ready - I particularly appreciate the parallel unit (or not really unit) test. I vote for fully deprecating the old `Mesh` constructors but will not fight if others think that should be done separately.

**Comment by pazner** [comment]
> `FiniteElementSpace::UpdateMeshPointer` makes me slightly nervous. This is an internal, special-purpose method. If it can't be made non-public, could we put it somewhere where it doesn't look like the user is OK to call it (i.e., _not_ close to FESpace::Update)? Otherwise I think the PR is fine.

@jakubcerveny, I made `FiniteElementSpace::UpdateMeshPointer` protected and made `Mesh::Swap` public and a friend function of `FiniteElementSpace`. Does that seem better to you?

**Comment by jakubcerveny** [comment]
That sounds good, thanks!

**Comment by v-dobrev** [comment]
> @jakubcerveny, I made `FiniteElementSpace::UpdateMeshPointer` protected and made `Mesh::Swap` public and a friend function of `FiniteElementSpace`. Does that seem better to you?

Was exposing `Mesh::Swap` as public necessary? I think there were some concerns with doing that back when it was added, so we kept it protected on purpose.

**Comment by pazner** [comment]
> Was exposing `Mesh::Swap` as public necessary? I think there were some concerns with doing that back when it was added, so we kept it protected on purpose.

It can't be a friend of `FiniteElementSpace` if it's protected. There are some "trickier" ways of achieving the same thing (creating a friend "wrapper" class), but I think that would be confusing.

**Comment by v-dobrev** [comment]
Can't we just make the whole class `Mesh` a friend of the FE space?

**Comment by pazner** [comment]
Yes, but I was hesitant to do that because it seems overkill. It is an option though.

Do you recall what the concerns about `Mesh::Swap` were?

**Comment by v-dobrev** [comment]
I think originally `Mesh::Swap` was not performing a complete swap of all data members. @jakubcerveny, do you remember any details? Looking at previous commits, I also see that before `bool non_geometry = false` was the default (now there is no default value), so maybe that's why we did not want to make it public.

**Comment by jakubcerveny** [comment]
I don't remember but what you're saying makes sense, maybe swapping Meshes was just too dangerous :-)

**Comment by v-dobrev** [comment]
One other potential issue is: what if someone tries to use it with a `ParMesh`? We may need to define `ParMesh::Swap` too. What do you think?

**Comment by pazner** [comment]
> One other potential issue is: what if someone tries to use it with a `ParMesh`? We may need to define `ParMesh::Swap` too. What do you think?

I already implemented `ParMesh::Swap`, but it was marked as protected. I made it public now, and marked `Mesh::Swap` to be protected in the `ParMesh` class (another option would be to make `Mesh::Swap` virtual, and then perform a check that the mesh to be swapped is really a `ParMesh` with `dynamic_cast` and ensure that `non_geometry` is true).

**Comment by v-dobrev** [comment]
Okay, I think this is ready to go.

One last thing: Will, do you want to add a bullet in CHANGLOG about the new `Mesh` constructor, `Mesh::MakeSimplicial`? It is one of the main things in the PR, so I think it is worth mentioning.

**Comment by v-dobrev** [comment]
LGTM

**Comment by v-dobrev** [comment]
LGTM

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
Merging in `next` has a number of conflicts.

I'll wait for #1817 to be merged in `master` and then merged here.

**Comment by v-dobrev** [comment]
@pazner, now that #1817 is merged in `master`, can you please merge `master` here to resolve conflicts/update code?

When that is ready, I hope the merge in `next` will be easy.

**Comment by v-dobrev** [comment]
LGTM

**Comment by v-dobrev** [comment]
Merged in `next` for testing ...

**Comment by tzanio** [comment]
Warning from tonight's regression testing:

```
adjoint_advection_diffusion.cpp(250): warning #1478: function "mfem::Mesh::Mesh(int, double)" (declared at line 767 of "../../mesh/mesh.hpp") was declared deprecated
     Mesh *mesh = new Mesh(mx+1, 2.);
                          ^
```

**Comment by v-dobrev** [comment]
Re-merged in `next`.

**Comment by tzanio** [comment]
@artv3, this can have an impact on your application. Can you give it a try and let us know if it works for you?

**Comment by v-dobrev** [comment]
ping: @artv3

**Comment by artv3** [comment]
I think it should be okay. I'll make a note to do more testing once we update mfem.

**Review by artv3 — APPROVED** [review]

**Pull request merged**
